common/src/main/java/io/netty/util/DirectCleaner.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util;
+
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.UnstableApi;
+
+import java.nio.ByteBuffer;
+
+@UnstableApi
+public final class DirectCleaner {
+
+    private static boolean ALLOW_CUSTOM_CLEANER = SystemPropertyUtil.getBoolean("io.netty.allowCustomCleaner", false);
+    private static volatile CustomCleaner customCleaner;
+
+    /**
+     * Try to deallocate the specified direct {@link ByteBuffer}. Please note this method does nothing if
+     * the current platform does not support this operation and no custom cleaner is provided or if the
+     * specified buffer is not a direct buffer.
+     */
+    public static void freeDirectBuffer(ByteBuffer buffer) {
+        // Copied locally to prevent race
+        CustomCleaner localCustomCleaner = DirectCleaner.customCleaner;
+        if (!ALLOW_CUSTOM_CLEANER || localCustomCleaner == null) {
```

Comment:
Can we check `!ALLOW_CUSTOM_CLEANER` before we fetch the volatile field ?

something like:

```
CustomCleaner localCustomCleaner;
if (!ALLOW_CUSTOM_CLEANER ||  (localCustomCleaner = DirectCleaner.customCleaner) == null)  {
    ...
} else {
    ....
}
```

Reply:
Yep. I was attempting to avoid that originally, but lost the change when I assigned the field locally. I did not think about assigning it in the `if` condition.